pkg/component/runtime/service.go

```diff
@@ -0,0 +1,441 @@
+			// TODO: switch to debug
+			s.log.Infof("observer checkin for %s service", cli)
+			sendExpected := false
+			changed := false
+			if s.state.State == client.UnitStateStarting {
+				// first observation after start set component to healthy
+				s.state.State = client.UnitStateHealthy
+				s.state.Message = fmt.Sprintf("Healthy: communicating with %s service", cli)
+				changed = true
+			}
+			if s.lastCheckin.IsZero() {
+				// first check-in
+				sendExpected = true
+			}
+			s.lastCheckin = time.Now().UTC()
+			if s.state.syncCheckin(checkin) {
+				changed = true
+			}
+			if s.state.unsettled() {
+				sendExpected = true
+			}
+			if sendExpected {
+				comm.CheckinExpected(s.state.toCheckinExpected())
+			}
+			if changed {
+				s.sendObserved()
+			}
+			if s.state.cleanupStopped() {
+				s.sendObserved()
+			}
+		}
+	}
+}
+
+// checkStatus checks service status, called on timer
+func (s *ServiceRuntime) checkStatus(cli service.Service, checkinPeriod time.Duration) {
+	status, err := cli.Status()
+	if err != nil {
+		s.forceCompState(client.UnitStateFailed, fmt.Sprintf("Failed: service status check error: %v", err))
+	}
+
+	if s.lastServiceStatus != status {
+		switch status {
+		case service.StatusUnknown:
+			s.forceCompState(client.UnitStateDegraded, fmt.Sprintf("Degraded: %s service status is unknown", cli))
+		case service.StatusStopped:
+			s.forceCompState(client.UnitStateStopped, fmt.Sprintf("Stopped: %s service", cli))
+		}
+	}
+
+	if status == service.StatusRunning {
+		now := time.Now().UTC()
+		if s.lastCheckin.IsZero() {
+			// never checked-in
+			s.missedCheckins++
+		} else if now.Sub(s.lastCheckin) > checkinPeriod {
+			// missed check-in during required period
+			s.missedCheckins++
+		} else if now.Sub(s.lastCheckin) <= checkinPeriod {
+			s.missedCheckins = 0
+		}
+		if s.missedCheckins == 0 {
+			s.compState(client.UnitStateHealthy, cli)
+		} else if s.missedCheckins > 0 && s.missedCheckins < maxCheckinMisses {
+			s.compState(client.UnitStateDegraded, cli)
+		} else if s.missedCheckins >= maxCheckinMisses {
+			// something is wrong; the service should be checking in
+			msg := fmt.Sprintf("Failed: %s service missed %d check-ins", cli, maxCheckinMisses)
+			s.forceCompState(client.UnitStateFailed, msg)
+		}
+	}
+	s.lastServiceStatus = status
+}
+
+// Watch returns the channel that sends component state.
+//
+// Channel should send a new state anytime a state for a unit or the whole component changes.
+func (s *ServiceRuntime) Watch() <-chan ComponentState {
+	return s.ch
+}
+
+// Start starts the service.
+//
+// Non-blocking and never returns an error.
+func (s *ServiceRuntime) Start() error {
+	s.actionCh <- serviceStart
+	return nil
+}
+
+// Update updates the currComp runtime with a new-revision for the component definition.
+//
+// Non-blocking and never returns an error.
+func (s *ServiceRuntime) Update(comp component.Component) error {
+	s.compCh <- comp
+	return nil
+}
```

Comment:
check ComponentRuntime. @blakerouse wrote the manager for the runtimes and the ComponentRuntime this way, I assume this the suitable approach for the ServiceRuntime.
@blakerouse Could you confirm if the non-buffered channels are ok here? Otherwise we would have to change this for ComponentRuntime as well.